### PR TITLE
Meta-model integration

### LINKED ISSUE #107: Re-train models
At the moment, we have 3 trained models:

- A UL tamper detection model (OCSVM)
- A SL tamper detection model (Catboost Classifier)
- A QoE SSIM predictor model (Catboost Regressor)

The UL and SL model are combined into a tamper detection meta model using an AND operator.

The trained models are accessible [here](https://storage.googleapis.com/verification-models/verification-metamodel.tar.xz).

The training code is on the [qoe_model_integration](https://github.com/livepeer/verification-classifier/tree/qoe_model_integration) branch.

These models are all trained using the same features extracted from the same [dataset](https://storage.googleapis.com/feature_dataset/data-qoe-large-2020-05-20.tar.xz) based on YT videos. The dataset contained renditions transcoded using the [ffmpeg CLI](https://github.com/livepeer/verification-classifier/blob/qoe_model_integration/scripts/encode_renditions.py#L59) which may not accurately reflect how renditions would be transcoded by LP orchestrators that use [LPMS](https://github.com/livepeer/lpms) (which is built on top of ffmpeg libraries). Furthermore, LP orchestrators could transcode renditions using either CPUs or Nvidia GPUs

Other differences with LP orchestrator transcoding include:

- Renditions can be transcoded using either CPUs or Nvidia GPUs
- Videos could have different frame rates besides 30fps

There may be other differences that are not mentioned above.

Given the possible differences between the renditions transcoded using the ffmpeg CLI used in the dataset that the models are currently trained on and the renditions that would be transcoded by LP orchestrators, it could be beneficial to re-train the models using an updated dataset that contains renditions transcoded using LP orchestrators with a variety of options that would be used in production.

We can generate a new feature dataset CSV file by:

- Collecting source videos
    - The dataset could consist of
        - Publicly available videos
            - YT (i.e. the existing dataset we use)
            - Netflix public data set mentioned [here](https://github.com/Netflix/vmaf/blob/master/resource/doc/datasets.md)
            - VQEG HD3 public data set mentioned [here](https://github.com/Netflix/vmaf/blob/master/resource/doc/datasets.md)
        - Videos being sent into the LP network
- Create renditions transcoded using options that a LP orchestrator would use (this [script](https://github.com/livepeer/lpms/tree/master/cmd/transcoding) might be useful)
- Generate attacks based on those renditions
- Extract features and dump them into the CSV file

We should make sure to re-train all 3 models using the same dataset.

One approach we could take is:

1. Create a CSV file with a few hundred rows
2. Train the models until the OCSVM achieves 95% TPR In training
3. Add more rows to the CSV file
4. Train the models with the updated CSV file

Before we do this we should do #98 to get a sense of how the current trained models perform when verifying renditions transcoded by GPU enabled LP orchestrators.

### PR DESCRIPTION
**Pull request by cyberj0g**
This resolves #107.
- Models re-trained for FPS-adjusted renditions
- Verification code refactored
- Merged FPS-aware VideoAssetProcessor and new VideoCapture
- API refactored
- Unit tests added

### CODE REVIEW

File: scripts/asset_processor/video_asset_processor.py [hunk 1]
```diff
@@ -23,630 +23,570 @@
...
-					if last_rendition_fps is not None and last_rendition_fps != fps:
-						raise Exception(f'Rendition has frame rate incompatible with other renditions: {fps}')
-					rendition['fps'] = fps
-					rendition['width'] = capture.width
-					rendition['height'] = capture.height
-				else:
-					raise Exception(f'Rendition not found: {path}')
-			finally:
-				capture.release()
+    """
```
Comment (yondonfu) [inline scripts/asset_processor/video_asset_processor.py:26]: Could we match the indentation of the file and then if indentation consistency is a problem in the repo we could update the indentation to be consistent across all files in a separate PR? This would make it easier to distinguish functional vs. formatting changes for review.

File: scripts/modeling_toolbox/utils.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    lowercase = lambda x: str(x).lower()
+    data_df.rename(lowercase, axis='columns', inplace=True)
+    data_df.drop(['dimension_x'], axis=1, inplace=True)
+    data_df.rename(columns={'attack':'rendition',
+                            'title':'source',
+                            'dimension': 'dimension_x'},
+                   inplace=True)
+    data_df['rendition'] = data_df['rendition'].apply(lambda x: set_rendition_name(x))
+    data_df['dimension_y'] = data_df['rendition'].apply(lambda x: int(x.replace('p','').split('_')[0]))
+    data_df['size_dimension_ratio'] = data_df['size'] / (data_df['dimension_y'] * data_df['dimension_x'])
```
Comment (yondonfu) [inline scripts/modeling_toolbox/utils.py:23]: I see elsewhere that `size_dimension_ratio` is calculated as `size / dimension`. Why do we calculate `size_dimension_ratio` as `size / dimension_y * dimension_x` here?

File: testing/tests/test_var_fps.py [hunk 3]
```diff
@@ -69,15 +62,22 @@ def test_classification(self):
                 rend_path = rendition_dir + os.path.sep + filename
                 if not os.path.exists(rend_path):
                     continue
-                res = self.verify_file(src, rend_path)
+                np.random.seed(123)
+                random.seed(123)
+                start = timeit.default_timer()
+                verification_result = verifier.verify(src, [{'uri': rend_path}])
+                tamper = float(verification_result[0]["tamper"])
```
Comment (yondonfu) [inline testing/tests/test_var_fps.py:69]: I think this will overwrite the `tamper` variable assigned on L60 which indicates if the rendition is actually tampered while `verification_result[0]["tamper"]` indicates if the meta model classified the rendition as tampered.

File: testing/tests/test_var_fps.py [hunk 4]
```diff
@@ -69,15 +62,22 @@ def test_classification(self):
...
+                res = {'score': tamper, 'score_ul': score_ul, 'score_sl': score_sl}
                 res['master_filename'] = filename
                 res['rendition_type'] = rendition_name
                 res['is_correct'] = not tamper
                 results.append(res)
         df_res: pd.DataFrame = pd.DataFrame(results)
         df_res.set_index(['master_filename', 'rendition_type'], inplace=True)
         df_res.sort_index(inplace=True)
-        df_res['prediction'] = df_res['score'] > 0
+        df_res['prediction'] = df_res['score'] <= 0
```
Comment (yondonfu) [inline testing/tests/test_var_fps.py:80]: Wouldn't this condition for the prediction make Tamper the positive class when we've been treating the Correct class as the positive class?

File: scripts/asset_processor/video_asset_processor.py (hunk 1 above)
Comment (cyberj0g) [reply scripts/asset_processor/video_asset_processor.py:26]: Indentation is consistent now, I believe. Would be great to also follow PEP8, and maybe add a test to check compliance. Actually, it's just that particular file I have converted to class, and then reformatted properly in a separate commit.

File: scripts/modeling_toolbox/utils.py (hunk 2 above)
Comment (cyberj0g) [reply scripts/modeling_toolbox/utils.py:23]: I think it's to better maintain feature scale for 2D features. Models are trained with features scaled like that now. Maybe @ndujar can comment.

File: testing/tests/test_var_fps.py (hunk 3 above)
Comment (cyberj0g) [reply testing/tests/test_var_fps.py]: Good catch, fixed.

File: testing/tests/test_var_fps.py (hunk 4 above)
Comment (cyberj0g) [reply testing/tests/test_var_fps.py]: Tamper is a positive class (positive score). In this test, it was vice versa. Fixed.

File: verifier/verifier.py [hunk 5]
```diff
@@ -1,276 +1,335 @@
...
+                # Add predictions to rendition dictionary
+                i = 0
+                for _, rendition in enumerate(renditions):
+                    if rendition['video_available']:
+                        rendition.pop('path', None)
+                        rendition['ssim_pred'] = float(predictions_df['ssim_pred'].iloc[i])
+                        rendition['ocsvm_dist'] = float(predictions_df['ocsvm_dist'].iloc[i])
+                        rendition['tamper_ul'] = int(predictions_df['ul_pred_tamper'].iloc[i])
+                        rendition['tamper_sl'] = int(predictions_df['sl_pred_tamper'].iloc[i])
+                        rendition['tamper'] = int(predictions_df['meta_pred_tamper'].iloc[i])
```
Comment (yondonfu) [inline verifier/verifier.py:208]: Perhaps we can name this key "tamper_meta" to match the name used in the API response

File: testing/tests/test_var_fps.py [hunk 6]
```diff
@@ -71,12 +71,12 @@ def test_classification(self):
...
                 res['master_filename'] = filename
                 res['rendition_type'] = rendition_name
-                res['is_correct'] = not tamper
+                res['is_tamper'] = tamper
                 results.append(res)
         df_res: pd.DataFrame = pd.DataFrame(results)
         df_res.set_index(['master_filename', 'rendition_type'], inplace=True)
         df_res.sort_index(inplace=True)
-        df_res['prediction'] = df_res['score'] <= 0
+        df_res['prediction'] = df_res['score'] > 0
```
Comment (yondonfu) [inline testing/tests/test_var_fps.py:79]: Isn't -1 = tamper? In that case, this would result in a False prediction when the meta model returns -1.

Comment (cyberj0g) [reply testing/tests/test_var_fps.py:79]: Negative score was "tamper" before, because OCSVM distance function, which has a negative value for an outlier, was used as a score. I think, it makes more sense to have Positive class designate abnormal (much less frequent) state, and model output in (0, 1) interval.

File: verifier/verifier.py (hunk 5 above)
Comment (cyberj0g) [reply verifier/verifier.py:208]: I think API (and even Verifier class) shouldn't expose much model implementation details (aside of debug purposes) to avoid confusion.

File: verifier/verifier.py [hunk 7]
```diff
@@ -1,276 +1,335 @@
...
+
+        self.loaded_scaler = load(open('{}/{}_{}.joblib'.format(self.model_dir,
+                                                                learning_type,
+                                                                scaler_type), 'rb'))
+        # Configure SL model for inference
+        model_name_sl = 'CB_Binary'
+        self.loaded_model_sl = CatBoostClassifier().load_model('{}/{}.cbm'.format(self.model_dir,
+                                                                                  model_name_sl))
+
+        # Configure SL model for inference
```
[the remark this answers is not in the document]
Comment (cyberj0g) [reply verifier/verifier.py]: QoE model is not used in classification currently, removed it.

[the remark this answers is not in the document]
Comment (yondonfu) [reply verifier/verifier.py]: We might want to eventually use the QoE model result to help evaluate transcoders on the network (after confirming that the transcoders have not tampered video). We're not doing that right now though so I think it's fine to remove the QoE model inference step for now and add it back in when we actually need the result.
